### DISCUSSION TITLE
on window integration alternative

### DISCUSSION THREAD

**Discussion by SanshooSenshi** [WindowTop/WindowTop-App#175] [Ideas]
If there is enough space on the window ( usualy it's always the case, except for many tabs on a browser ) 
It could be a better workflow if we could have shortcuts displaying instead of waiting for the little "arrow" to show then click on it ... 

with a on " mouse over " = no transparency for each options ...
![windowTop integration](https://user-images.githubusercontent.com/85396423/146167448-26a9c309-ad1f-4ff2-bc3a-4206ace7a608.png)

**Comment by gileli121** [comment]
Thanks for the idea. It is an old one.
Many developers tried to make it done but no one was able to do it right.
In deskpins for example the extra button was above the window title.
The hard challenge is how to make it done right so that the extra button will not be above the window title

Anyway, I will try to overcome this challenge and implement this feature

**Comment by gileli121** [comment]
I am investigating it.
I wonder how google added this button:
![image](https://user-images.githubusercontent.com/17680514/149384150-fdf6375d-b0ee-4168-8493-655206bab209.png)

**Reply by cpasoft** [reply]
This would be great!

**Comment by gileli121** [comment]
Created task 
https://github.com/BiGilSoft/WindowTop/issues/192

**Comment by gileli121** [comment]
I finished the investigation.
It is possible to put an extra button on the window here:
![image](https://user-images.githubusercontent.com/17680514/149452688-685974f8-d882-4992-ba87-ee1bcdbec09f.png)

This is not the problem.
The problem is how to avoid the button will overlap the other stuff.
In regular windows, Windows put `...` in this case so you never see that the default buttons will overlap the text
![image](https://user-images.githubusercontent.com/17680514/149452910-b80aa5d4-6d7f-4567-b094-4414fb5d75bc.png)

I found that for regular windows I can fake the same effect by trimming the text myself and making a call to [SetWindowText](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowtexta) 

So for regular windows, the problem can be solved.

But for windows that do not show window title next to these buttons, this is a problem.
Examples of such windows are chrome.

Chrome shows tabs next to these buttons. How do you trim tabs? That's the question... No idea. It looks like chrome draws everything manually, including the minimize, maximize, close buttons. 


By checking the user request, it seems that you are fine with such an issue... But I am not. 
If there is no way to save this extra space for the button, then it will look broken. Maybe it is an option for the regular windows that show only title.

**Comment by gileli121** [comment]
I have another proposal:

1. If the window is maximized, put the toolbar here in the taskbar:
![image](https://user-images.githubusercontent.com/17680514/149454525-a7c357c0-e7af-42dc-a246-7c706b4c6d76.png)

2. If the window is not maximized, put the toolbar here:
![image](https://user-images.githubusercontent.com/17680514/149454840-dcf3f8c6-d54a-44b9-9379-a459a8a3cb48.png)

But the toolbar theme/style should be more minimal so it will not look ugly.

This is the idea anyway

**Comment by gileli121** [comment]
But I wonder why you are asking and voting for this feature...
Do you used to click on the arrow button here 
![image](https://user-images.githubusercontent.com/17680514/149456076-e81e2b1e-499a-431a-a6cb-cdce04fc9dd6.png)
?
And because you think that you should click on it, are you asking for a way to avoid this extra click?

If this is the case, your reason is invalid... You don't need to click on the arrow...
When the mouse is above the arrow, the toolbar will show up after 0.3-0.4 seconds.

It looks easy enough process so I can't see valid reason

**Comment by cpasoft** [comment]
I don't like the taskbar idea... I like the arrow as it is right now...

I cannot find problem with it... It's working very well...

**Comment by gileli121** [comment]
I also don't like.
I think that the existing solution is better anyway.
And I did not planed to remove the existing solution anyway. Just to add another option.

I think that I will resolve it as won't-do

**Comment by gileli121** [comment]
@cpasoft 
I have an idea here. But it will wait after #194 (or maybe not)..

If the mouse is above this button:
![image](https://user-images.githubusercontent.com/17680514/153253362-b3594188-c7f6-4989-a8d7-c0100027f64d.png)
Then after 0.5 seconds (or less?) show this:
![image](https://user-images.githubusercontent.com/17680514/153253548-5f821597-5a04-4bc3-bc40-08d3c280da3e.png)


What do you think?
If this mode is enabled, It will replace the arrow button

And thanks to https://github.com/microsoft/PowerToys/issues/16197 that helped me to get this idea

**Reply by cpasoft** [reply]
> @cpasoft I have an idea here. But it will wait after #194 (or maybe not)..
> 
> If the mouse is above this button: ![image](https://user-images.githubusercontent.com/17680514/153253362-b3594188-c7f6-4989-a8d7-c0100027f64d.png) Then after 0.5 seconds (or less?) show this: ![image](https://user-images.githubusercontent.com/17680514/153253548-5f821597-5a04-4bc3-bc40-08d3c280da3e.png)
> 
> What do you think? If this mode is enabled, It will replace the arrow button

It's an smart feature... I like it !!!...
I think this would be very useful... I vote for it!!!

One question about this.... It would be available on Windows 10, or only on Windows 11??

**Reply by gileli121** [reply]
Great! I will implement it on the weekend.
Yes, it will work on Windows 10 also.

**Comment by cpasoft** [comment]
Another question... What will happen with this windows without the minimize symbol??

**Reply by gileli121** [reply]
@cpasoft 
In Windows, these cases are possible:

Case 1 - Only close button
![image](https://user-images.githubusercontent.com/17680514/153304584-38591d47-91eb-4ec5-a81c-79c5e4767a12.png)

Case 2 - Minimize button disabled
![image](https://user-images.githubusercontent.com/17680514/153304179-058b8aa6-2b48-425a-9890-31bd2d00777d.png)

Case 3 - Maximize button disabled
![image](https://user-images.githubusercontent.com/17680514/153304884-294c8378-263b-43af-b0bb-a0df34e4e2e7.png)

(If you will try to disable the maximize and minimize, the OS will not let you.. it will just remove them so you end with case 1)

Case 4 - No buttons at all
![image](https://user-images.githubusercontent.com/17680514/153305061-36339090-579f-4abf-b7d7-260d602a6a0e.png)


I think that when the minimize button disabled or not there, I should show the regular arrow.

**Comment by gileli121** [comment]
@cpasoft 
There is small issue. the arrow have right click menu:
![image](https://user-images.githubusercontent.com/17680514/153308489-6cbac61e-34af-4861-81a9-8e4fb3d2e9cc.png)

I can't do it with the minimize button..
So I think that the toolbar will have somehow to include this menu also but I don't know how yet

**Reply by cpasoft** [reply]
> @cpasoft There is small issue. the arrow have right click menu: ![image](https://user-images.githubusercontent.com/17680514/153308489-6cbac61e-34af-4861-81a9-8e4fb3d2e9cc.png)
> 
> I can't do it with the minimize button.. So I think that the toolbar will have somehow to include this menu also but I don't know how yet

You could move the right click actions to another button or to the config...

**Comment by gileli121** [comment]
@cpasoft 
See
https://github.com/BiGilSoft/WindowTop/issues/192#issuecomment-1037260917
